internal/cluster/distribution/pke/pkeaws/image_selector.go

```diff
@@ -0,0 +1,133 @@
+// Copyright © 2020 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pkeaws
+
+import (
+	"context"
+	"sync"
+
+	"emperror.dev/errors"
+)
+
+// +testify:mock:testOnly=true
+
+// ImageSelector chooses an image based on the selection criteria.
+// It returns an ImageNotFoundError when no images can be found matching the provided criteria.
+type ImageSelector interface {
+	SelectImage(ctx context.Context, criteria ImageSelectionCriteria) (string, error)
+}
+
+// ImageSelectionCriteria contains all parameters for selecting an image.
+type ImageSelectionCriteria struct {
+	Region            string
+	InstanceType      string
+	PKEVersion        string
+	KubernetesVersion string
+	OperatingSystem   string
+	ContainerRuntime  string
+}
+
+// ImageNotFoundError is returned by an ImageSelector when it cannot find an image matching the provided criteria.
+const ImageNotFoundError = errors.Sentinel("no images found matching the selection criteria")
+
+// RegionMapImageSelector selects an image based on the region in the selection criteria.
+type RegionMapImageSelector map[string]string
+
+func (r RegionMapImageSelector) SelectImage(_ context.Context, criteria ImageSelectionCriteria) (string, error) {
+	image, ok := r[criteria.Region]
+	if !ok {
+		return "", ImageNotFoundError
+	}
+
+	return image, nil
+}
+
+// ImageSelectorChain tries to select an image using a number of selectors.
+// When one fails, it moves onto the next in the chain.
+type ImageSelectorChain struct {
+	imageSelectors []ImageSelector
+	names          []string
+
+	mu sync.RWMutex
+
+	logger     Logger
+	loggerOnce sync.Once
+
+	errorHandler     ErrorHandler
+	errorHandlerOnce sync.Once
+}
+
+// NewImageSelectorChain returns a new ImageSelectorChain.
+func NewImageSelectorChain(logger Logger, errorHandler ErrorHandler) ImageSelectorChain {
+	return ImageSelectorChain{
+		logger:       logger,
+		errorHandler: errorHandler,
+	}
+}
+
+func (s *ImageSelectorChain) getLogger() Logger {
+	s.loggerOnce.Do(func() {
+		if s.logger == nil {
+			s.logger = NoopLogger{}
+		}
+	})
```

Comment:
**Question**: what is the pro for initializing a NoopLogger here as opposed to the constructor (by checking the function argument for `nil`) where no synchronization would be required because only the constructor itself has reference to the newly created object?

(Same for https://github.com/banzaicloud/pipeline/pull/3049/files#diff-93108070aa9a2926c7ddda88735708c4R95)

Reply:
It's possible to use the struct without the constructor, if you don't want to specify a logger and an error handler.

Reply:
I even considered using public fields for the logger and the error handler, but I settled with public fields and a constructor for now. We can always change it (in a backward compatible manner) later.